R/tokens_xptr.R

```diff
@@ -0,0 +1,209 @@
+    build_tokens(result, 
+                 types = NULL, 
+                 padding = TRUE, 
+                 docvars = attrs[["docvars"]], 
+                 meta = attrs[["meta"]], 
+                 class = attrs[["class"]])
+}
+
+#' @rdname tokens_xptr
+#' @method as.tokens_xptr tokens_xptr
+#' @export
+as.tokens_xptr.tokens_xptr <- function(x) {
+    attrs <- attributes(x)
+    result <- cpp_copy_xptr(x)
+    rebuild_tokens(result, attrs)
+}
+
+#' @export
+ndoc.tokens_xptr <- function(x) {
+    cpp_ndoc(x)
+}
+
+#' @export
+types.tokens_xptr <- function(x) {
+    cpp_get_types(x, TRUE)
+}
+
+#' @export
+ntoken.tokens_xptr <- function(x, ...) {
+    structure(cpp_ntoken(x), names = docnames(x))
+}
+
+# #' @export
+# docnames.tokens_xptr <- function(x) {
+#     get_docvars(x, "docname_", FALSE, TRUE, TRUE)
+# }
+# 
+# #' @export
+# docid.tokens_xptr <- function(x) {
+#     get_docvars(x, "docid_", FALSE, TRUE, TRUE)
+# }
+# 
+# #' @export
+# segid.tokens_xptr <- function(x) {
+#     get_docvars(x, "segid_", FALSE, TRUE, TRUE)
+# }
+
+#' @method as.tokens tokens_xptr
+#' @export
+as.tokens.tokens_xptr <- function(x, ...) {
+    attrs <- attributes(x)
+    result <- cpp_as_list(x)
+    build_tokens(result, 
+                 types = attr(result, "types"), 
+                 padding = TRUE, 
+                 docvars = attrs[["docvars"]], 
+                 meta = attrs[["meta"]],
+                 class = attrs[["class"]])
+}
+
+#' @method as.list tokens_xptr
+#' @export
+as.list.tokens_xptr <- function(x, ...) {
+    as.list(as.tokens(x), ...)
+}
+
+#' @method [ tokens_xptr
+#' @export
+"[.tokens_xptr" <- function(x, i, drop_docid = TRUE) {
+    if (missing(i)) return(x)
+    attrs <- attributes(x)
+    
+    index <- seq_along(docnames(x))
+    names(index) <- docnames(x)
+    index <- index[i]
+    
+    if (any(is.na(index)))
+        stop("Subscript out of bounds")
+    
+    build_tokens(
+        cpp_subset(x, index),
+        attrs[["types"]],
+        docvars = reshape_docvars(attrs[["docvars"]], index, drop_docid = drop_docid),
+        meta = attrs[["meta"]],
+        class = attrs[["class"]]
+    )
+}
+
+#' @method head tokens_xptr
+#' @export
+head.tokens_xptr <- function(x, n = 6L, ...) {
+    x[head(seq_len(ndoc(x), ...), n)]
+}
+
+#' @method tail tokens_xptr
+#' @export
+tail.tokens_xptr <- function(x, n = 6L, ...) {
+    x[tail(seq_len(ndoc(x), ...), n)]
+}
+
+
+#' @export
+tokens_subset.tokens_xptr <- function(x, subset, drop_docid = TRUE, ...) {
+    
+    check_dots(...)
+    
+    attrs <- attributes(x)
+    docvar <- get_docvars(x, user = TRUE, system = TRUE)
+    r <- if (missing(subset)) {
+        rep_len(TRUE, ndoc(x))
+    } else {
+        e <- substitute(subset)
+        r <- eval(e, docvar, parent.frame())
+        r & !is.na(r)
+    }
+    return(x[r, drop_docid = drop_docid])
+}
+
+#' @method dfm tokens_xptr
+#' @export
+dfm.tokens_xptr <- function(x, tolower = TRUE, remove_padding = FALSE, ...) {
```

Comment:
👍 👍 👍 This is a real game changer, since compiling a dfm for a large tokens object can be hugely time-consuming.

Reply:
It is the time to defunct all the Swiss Army arguments of `dfm()`.

Reply:
Agreed but let's leave all deprecations and defuncting to a separate PR.